src/plugins/share/public/types.ts

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import {
+  EuiContextMenuPanelDescriptor,
+  EuiContextMenuPanelItemDescriptor,
+} from '@elastic/eui/src/components/context_menu/context_menu';
+
+/**
+ * @public
+ * Properties of the current object to share. Registered share
+ * menu providers will provide suitable items which have to
+ * be rendered in an appropriate place by the caller.
+ *
+ * It is possible to use the static function `showShareContextMenu`
+ * to render the menu as a popover.
+ * */
+export interface ShareMenuItemProps {
+  objectType: string;
+  objectId?: string;
+  /**
+   * Current url for sharing. This can be set in cases where `window.location.href`
+   * does not contain a shareable URL (e.g. if using session storage to store the current
+   * app state is enabled). In these cases the property should contain the URL in a
+   * format which makes it possible to use it without having access to any other state
+   * like the current session.
+   *
+   * If not set it will default to `window.location.href`
+   */
+  shareableUrl: string;
+  sharingData: unknown;
+  isDirty: boolean;
+  onClose: () => void;
+}
+
+/**
+ * @public
+ * Eui context menu entry shown directly in the context menu. `sortOrder` is
+ * used to order the individual items in a flat list returned by all registered
+ * menu providers.
+ * */
+export interface ShareContextMenuPanelItem extends EuiContextMenuPanelItemDescriptor {
+  sortOrder: number;
+}
+
+/**
+ * @public
+ * Definition of a menu item rendered in the share menu. `shareMenuItem` is shown
+ * directly in the context menu. If the item is clicked, the `panel` is shown.
+ * */
+export interface ShareMenuItem {
+  shareMenuItem: ShareContextMenuPanelItem;
+  panel: EuiContextMenuPanelDescriptor;
+}
+
+/** @public */
+export interface ShareMenuProvider {
+  readonly id: string;
+
+  getShareMenuItems: (actionProps: ShareMenuItemProps) => ShareMenuItem[];
+}
+
+/** @public */
+export interface ShowProps extends Omit<ShareMenuItemProps, 'onClose'> {
+  anchorElement: HTMLElement;
+  allowEmbed: boolean;
+  allowShortUrl: boolean;
+}
```

Comment:
I forgot to export them from the top level. I think we should export these because they are indirectly part of the public api - `ShareContextMenuPanelItem` is contained within `ShareMenuItem` which is returned by the `ShareMenuProvider` consumers can register and `ShowProps` is the parameter of the public `showShareContextMenu` function.

We don't need to export them necessarily but in the past I always found it helpful to have access to these types to build the objects for the api in a more flexible way in my own code. Like this:
```ts
function generateTheShowPropsPlease(): ShowProps {
 // Without being able to explicitly declare the return type here,
 // I won't have autosuggestions for the keys of the return object.
 return { /* ... */ };
}


// ... same other things, maybe in a nested function

share.showShareContextMenu(generateTheShowPropsPlease());

```

Reply:
👍  I'm in support of exporting them, but I might propose renaming `ShowProps`. I know you didn't choose this name, but since we are already touching all usages of it in this PR it would be a good time to update it.

It'd be nice to have a name that's more specific. Something like `ShowShareContextMenuProps` is a mouthful, but IMO this or something similar would be an improvement.

Reply:
You are absolutely right, I changed some of the names:
* `ShowProps` becomes `ShowShareMenuOptions` - props is kind of unusual outside of react anyway IMHO
* `ShareMenuItemProps` becomes `ShareContext` because it's basically the usage context we are querying share menu items for
* `showShareContextMenu` becomes `toggleShareContextMenu` because that's what it does.